File: x10.runtime.cws/src/x10/runtime/cws/Pool.java

/*
 * Written by Doug Lea with assistance from members of JCP JSR-166
 * Expert Group and released to the public domain, as explained at
 * http://creativecommons.org/licenses/publicdomain
 */
package x10.runtime.cws;

import java.util.concurrent.*;
import java.util.concurrent.locks.*;

import static x10.runtime.cws.ClosureStatus.*;




/**
 * The host and external interface for ForkJoinTasks. A ForkJoinPool
 * manages a group of specialized worker threads that perform
 * ForkJoinTasks. It also provides the entry point for tasks submitted
 * from non-ForkJoinTasks.
 *
 * <p> Class ForkJoinPool does not implement the ExecutorService
 * interface because it only executes ForkJoinTasks, not arbitrary
 * Runnables. However, for the sake of uniformity, it supports all
 * ExecutorService lifecycle control methods (such as shutdown).
 */
public class Pool {
    /*
     * This is an overhauled version of the framework described in "A
     * Java Fork/Join Framework" by Doug Lea, in, Proceedings, ACM
     * JavaGrande Conference, June 2000
     * (http://gee.cs.oswego.edu/dl/papers/fj.pdf). It retains most of
     * the basic structure, but includes a number of algorithmic
     * improvements, along with integration with other
     * java.util.concurrent components.
     */

    /**
     * Main lock protecting access to threads and run state. You might
     * think that having a single lock and condition here wouldn't
     * work so well. But serializing the starting and stopping of
     * worker threads (its main purpose) helps enough in controlling
     * startup effects, contention vs dynamic compilation, etc to be
     * better than alternatives.
     */
    private final ReentrantLock lock = new ReentrantLock();

    /**
     * Condition triggered when new work is available so workers
     * should awaken if blocked. Also used for timed waits when work
     * is known to be available but repeated steal attempts fail
     */
    private final Condition work = lock.newCondition();

    /**
     * The amount of time to block on "work" condition when a thread
     * repeatedly fails to be able to get or steal a task yet
     * activeJobs is nonzero (thus indicating that there might
     * again be work available in the future). As a rough guide,
     * this should be around twice the time for a full park/unpark
     * context switch.
     */
    private static final long IDLE_SLEEP_NANOS = 100L * 1000 * 1000;

    /**
     * The number of times a worker should call getJob, yielding on
     * failure, before sleeping . This value is a compromise between
     * responsiveness and good citizenship. Too small leads to
     * needless sleeps while other threads are just momentarily
     * delayed in GC etc.  Too big leads to CPU wastage
     */
    private static final int YIELDS_BEFORE_SLEEP = 64;

    /**
     * Tracks whether pool is running, shutdown, etc. Modified
     * only under lock, but volatile to allow concurrent reads.
     */
    private volatile int runState;

    // Values for runState
    static final int RUNNING    = 0;
    static final int SHUTDOWN   = 1;
    static final int STOP       = 2;
    static final int TERMINATED = 3;

    /**
     * The pool of threads. Currently, all threads are created
     * upon construction. However, all usages of workers array
     * are prepared to see null entries, allowing alternative
     * schemes in which workers are added more lazily.
     */
    private final Worker[] workers;

    
    /**
     * The number of jobs that are currently executing in the pool.
     * This does not include those jobs that are still in job queue
     * waiting to be taken by workers.
     */
    private int activeJobs = 0;

    /** The number of workers that have not yet terminated */
    private int runningWorkers = 0;

    /**
     * Condition for awaitTermination. Triggered when
     * activeWorkers reaches zero.
     */
    private final Condition termination = lock.newCondition();

    /**
     * The uncaught exception handler used when any worker
     * abrupty terminates
     */
    private volatile Thread.UncaughtExceptionHandler ueh;

    /**
     * Creates a ForkJoinPool with a pool size equal to the number of
     * processors available on the system.
     */
    public Pool() {
        this(Runtime.getRuntime().availableProcessors());
    }

    /**
     * Creates a ForkJoinPool with the indicated number
     * of Worker threads.
     */
    public Pool(int poolSize) {
        if (poolSize <= 0) throw new IllegalArgumentException();
        Worker.workers = workers = new Worker[poolSize];
        lock.lock();
        try {
            for (int i = 0; i < poolSize; ++i) {
                Worker r = new Worker(this, i);
                workers[i] = r;
               
            }
            for (int i = 0; i < poolSize; ++i) {
            	workers[i].start();
            	++runningWorkers;
            }
            
        } finally {
            lock.unlock();
        }
    }
  
    /**
     * Sets the handler for internal worker threads that terminate due
     * to uncaught Errors or other unchecked Throwables encountered
     * while executing tasks. Since such errors are not in generable
     * recoverable, they are not managed by ForkJoinTasks themselves,
     * but instead cause threads to die, invoking the handler, and
     * then being replaced.  In many usage contexts, it is probably a
     * better ideas to install a handler to either shutdown the pool
     * or exit the program when such errors occur.  Unless set, the
     * current default or ThreadGroup handler is used as handler.
     * @param h the new handler
     * @return the old handler, or null if none
     */
    public Thread.UncaughtExceptionHandler setUncaughtExceptionHandler(Thread.UncaughtExceptionHandler h) {
        final ReentrantLock lock = this.lock;
        Thread.UncaughtExceptionHandler old = null;
        lock.lock();
        try {
            old = ueh;
            ueh = h;
            for (int i = 0; i < workers.length; ++i) {
                Worker w = workers[i];
                if (w != null)
                    w.setUncaughtExceptionHandler(h);
            }
        } finally {
            lock.unlock();
        }
        return old;
    }

    /**
     * Returns the number of worker threads in this pool.
     *
     * @return the number of worker threads in this pool
     */
    public int getPoolSize() {
        return workers.length;
    }

    /**
     * Returns <tt>true</tt> if this pool has been shut down.
     *
     * @return <tt>true</tt> if this pool has been shut down
     */
    public boolean isShutdown() {
        return runState >= SHUTDOWN;
    }

    /**
     * Returns <tt>true</tt> if all tasks have completed following shut down.
     * Note that <tt>isTerminated</tt> is never <tt>true</tt> unless
     * either <tt>shutdown</tt> or <tt>shutdownNow</tt> was called first.
     *
     * @return <tt>true</tt> if all tasks have completed following shut down
     */
    public boolean isTerminated() {
        return runState == TERMINATED;
    }

    /**
     * Initiates an orderly shutdown in which previously submitted
     * tasks are executed, but no new tasks will be accepted.
     * Invocation has no additional effect if already shut down.
     */
    public void shutdown() {
        // todo security checks??
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (runState < SHUTDOWN) {
                
                    runState = SHUTDOWN;
                
            }
        } finally {
            lock.unlock();
        }
    }

    /**
     * Attempts to stop all actively executing tasks, and halts the
     * processing of waiting tasks.
     */
    public void shutdownNow() {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
           // tryTerminate();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Blocks until all tasks have completed execution after a shutdown
     * request, or the timeout occurs, or the current thread is
     * interrupted, whichever happens first.
     *
     * @param timeout the maximum time to wait
     * @param unit the time unit of the timeout argument
     * @return <tt>true</tt> if this executor terminated and
     *         <tt>false</tt> if the timeout elapsed before termination
     * @throws InterruptedException if interrupted while waiting
     */
    public boolean awaitTermination(long timeout, TimeUnit unit)
        throws InterruptedException {
        long nanos = unit.toNanos(timeout);
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            for (;;) {
                if (runState == TERMINATED)
                    return true;
                if (nanos <= 0)
                    return false;
                nanos = termination.awaitNanos(nanos);
            }
        } finally {
            lock.unlock();
        }
    }

    // Internal methods that may be invoked by workers

    final boolean isStopped() {
        return runState >= STOP;
    }

    /**
     * Return the workers array; needed for random-steal by Workers.
     */
    final Worker[] getWorkers() {
        return workers;
    }
    
    public  void submit(Job job) {
        addJob(job);
       
    }

    /**
     * Enqueue an externally submitted task
     */
    private void addJob(Job job) {
        final ReentrantLock lock = this.lock;
        boolean ok;
        lock.lock();
        try {
            if (ok = (runState == RUNNING)) {
                jobs.add(job);
                work.signalAll();
            }
        } finally {
            lock.unlock();
        }
        if (!ok)
            throw new RejectedExecutionException();
    }

   
    /**
     * Termination callback from dying worker.
     */
    final void workerTerminated(Worker r, int index) {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (runState >= STOP) {
                if (--runningWorkers <= 0) {
                    runState = TERMINATED;
                    termination.signalAll();
                }
            }
            else if (index >= 0 && index < workers.length &&
                     workers[index] == r) {
                Worker replacement = new Worker(this, index);
                if (ueh != null)
                    replacement.setUncaughtExceptionHandler(ueh);
                workers[index] = replacement;
                replacement.start();
            }
        } finally {
            lock.unlock();
        }
    }
    
    final void jobCompleted() {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (--activeJobs <= 0 && runState == SHUTDOWN && jobs.isEmpty())
                tryTerminate();
        } finally {
            lock.unlock();
        }
    }
    public void tryTerminate() {
    	// do nothing.
    }
    
    /**
     * External submission queue.  "Jobs" are tasks submitted to the
     * pool, not internally generated.
     */
    private final JobQueue jobs = new JobQueue();
    /**
     * Returns a job to run, or null if none available.
     * @param yields number of times caller has repeatedly failed to
     * find tasks. Upon threshold, sleeps a while unless woken by some
     * other thread that finds work.
     */
    final Closure getJob(int yields) {
        Closure task = null;
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            boolean timeout = false;
            task = jobs.poll();
            if (task == null) {
                if (activeJobs == 0)
                    work.await();
                else if (yields >= YIELDS_BEFORE_SLEEP)
                    timeout = work.awaitNanos(IDLE_SLEEP_NANOS) <= 0;
                else
                    timeout = true;
                if (!timeout)
                    task = jobs.poll();
            }
            if (task != null)
                ++activeJobs;
            if (task != null || (!timeout && activeJobs > 0))
                work.signal();
        } catch (InterruptedException ie) { 
            // ignore/swallow 
        } finally {
            lock.unlock();
        }
        return task;
    }
    /**
     * A JobQueue is a simple array-based circular queue.
     * Basically a stripped-down variant of ArrayDeque
     */
    static final class JobQueue {
        static final int INITIAL_JOBQUEUE_CAPACITY = 64;
        Job[] elements = (Job[]) new Job[INITIAL_JOBQUEUE_CAPACITY];
        int head;
        int tail;

        boolean isEmpty() {
            return head == tail;
        }

        void add(Job e) {
            elements[tail] = e;
            if ( (tail = (tail + 1) & (elements.length - 1)) == head)
                doubleCapacity();
        }

        Job poll() {
            int h = head;
            Job result = elements[h]; 
            if (result != null) {
                elements[h] = null;
                head = (h + 1) & (elements.length - 1);
            }
            return result;
        }
        
        void doubleCapacity() {
            int p = head;
            int n = elements.length;
            int r = n - p; 
            int newCapacity = n << 1;
            if (newCapacity < 0)
                throw new IllegalStateException("Job queue capacity exceeded");
            Job[] a = (Job[]) new Job[newCapacity];
            System.arraycopy(elements, p, a, 0, r);
            System.arraycopy(elements, 0, a, r, p);
            elements = a;
            head = 0;
            tail = n;
        }
    }
  
}

File: tags/SF_RELEASE_1_7_3/x10.lib/src/sched/Pool.cc
/*
 * (c) Copyright IBM Corporation 2007
 *
 * $Id: Pool.cc,v 1.21 2007-12-26 12:53:11 srkodali Exp $
 * This file is part of X10 Runtime System.
 */

#include <x10/xws/Pool.h>
#include <x10/xws/ActiveWorkerCount.h>
#include <x10/xws/Runnable.h>
#include <x10/xws/Job.h>
#include <x10/xws/Lock.h>
#include <x10/xws/Worker.h>
#include <x10/xws/Sys.h>
#include <x10/xassert.h>
#include <stdlib.h>
#include <strings.h>
#include <iostream>

using namespace x10lib_xws;
using namespace std;


Pool::~Pool() {
	int i;

	for(i=0; i<workers.size(); i++) {
	  if(workers[i] != NULL) {
	    pthread_join(id[i], (void **)NULL);
	    free(workers[i]->area());
	  }
	}
	free(id);

 	for(i=0;i<workers.size();i++)
 	  delete workers[i];
	workers.clear();
	pthread_cond_destroy(&work);
	pthread_cond_destroy(&termination);
	delete lock_var;
	delete barrier;
	delete jobs;
}


void Pool::callBackFunc::each_thread(Pool *p,int d)
{
	 Worker *ws = new Worker(d, p);
	assert(ws != NULL);
	 p->workers[d]=ws;
	 MEM_BARRIER();
     //barrier->barrier();
     /*if (id == 0)
	  ws->run(invoke_main);
     else
      ws->run((Closure *) NULL);*/
    ws->run();
    //cerr<<id<<":: returned from Worker::run()"<<endl;
}

void *Pool::each_thread_wrapper(void *arg) {
  assert(arg != NULL);
	callBackFunc *cbf = (callBackFunc *) arg;
	assert(cbf!=NULL);
	cbf->each_thread(cbf->cl, cbf->id);
	delete cbf;
	return NULL;
}


// Pool::Pool() { }

/*Anonymous innner class argument to ActiveWorkerCount from Java code*/
//class anon_Runnable : public virtual Runnable {
class anon_Runnable : public Runnable {
private:
  Pool *p;
  ~anon_Runnable(){} //cannot inherit
public:
  anon_Runnable(Pool *p) {
    this->p = p;
  }
  //virtual void run() {
  void run() {
    Job *job = (Job *) p->currentJob;
    if (job != NULL && 
	job->requiresGlobalQuiescence()) {
       job->completed();
       job->jobCompleted();
    }
    p->currentJob = NULL;
  }
};

Pool::Pool(int numThreads) {
	
	long i;
	int res;
	pthread_attr_t attr;
	
	if (numThreads <=0) { cout << "Illegal argument"; abort(); }

	currentJob = NULL;
	
	lock_var = new PosixLock();
	assert(lock_var != NULL);

	pthread_cond_init (&work, NULL);
	pthread_cond_init (&termination, NULL);

	barrier = new ActiveWorkerCount(new anon_Runnable(this));
	assert(barrier!= NULL);

	num_workers = numThreads;
	runningWorkers = 0;
	runState = 0;
	activeJobs = 0;
	activeOnJobAtomic = 0;
	jobs =  new JobQueue();
	assert(jobs != NULL);
	joinCount = 0;

	workers.resize(numThreads);
	for(i=0; i<workers.size(); i++) {
	  workers[i] = NULL;
	}

	  
	id = (pthread_t *)malloc(num_workers * sizeof(pthread_t));
	assert(id != NULL);

	pthread_attr_init(&attr); 
	pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM); 
	 
	lock_var -> lock_wait_posix();

	callBackFunc *ptToFunc;
	  
	for (i = 0; i < num_workers; i++)
	{
	  ptToFunc = new callBackFunc(this, i);
	  assert(ptToFunc != NULL);
// 	  ptToFunc.id = i;
// 	  ptToFunc.cl = this;
	  res = pthread_create(id + i, 
			       &attr,
			       Pool::each_thread_wrapper, 
			       (void *) ptToFunc);
	  if (res) {cout << "could not create"; abort(); }
	  runningWorkers++;
	  
	}
	lock_var -> lock_signal_posix();
}

/*
Pool::shutDown() {

	for (i = 0; i < num_workers; i++)
	{
		  res = pthread_join(id[i], NULL);
		  if (res) cout << "could not join";
	}
}*/


/**
     * Sets the handler for internal worker threads that terminate due
     * to uncaught Errors or other unchecked Throwables encountered
     * while executing tasks. Since such errors are not in generable
     * recoverable, they are not managed by ForkJoinTasks themselves,
     * but instead cause threads to die, invoking the handler, and
     * then being replaced.  In many usage contexts, it is probably a
     * better ideas to install a handler to either shutdown the pool
     * or exit the program when such errors occur.  Unless set, the
     * current default or ThreadGroup handler is used as handler.
     * @param h the new handler
     * @return the old handler, or null if none
     */
/*
UncaughtExceptionHandler *Pool::setUncaughtExceptionHandler(UncaughtExceptionHandler *h) {
        UncaughtExceptionHandler *old = NULL;
				abort(); // TODO RAJ
        const PosixLock *l = this->lock_var;
        l->lock_wait_posix();
        old = ueh;
        ueh = h;
        for (int i = 0; i < workers.size(); ++i) {
        	Worker *w = workers[i];
            if (w != NULL)
                    w.setUncaughtExceptionHandler(h);
            
        }
        l->lock_signal_posix();
        return old;
}
*/

int Pool::getPoolSize() const {
        return workers.size();
}
int Pool::activeOnJob() const {
    	return activeOnJobAtomic;
}

bool Pool::isShutdown() volatile {
        return runState >= SHUTDOWN;
}

bool Pool::isTerminated() const {
        return runState == TERMINATED;
}

void Pool::initFrameGenerator(FrameGenerator *fg) {
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            t->setFrameGenerator(fg);
        }
}

   /**
     * Initiates an orderly shutdown in which previously submitted
     * tasks are executed, but no new tasks will be accepted.
     * Invocation has no additional effect if already shut down.
     */
void Pool::shutdown() {
        // todo security checks??
  PosixLock *l = this->lock_var;
  l->lock_wait_posix();

  if (runState < SHUTDOWN) {
    runState = SHUTDOWN;
	  MEM_BARRIER(); // for volatile update -- TODO RAJ
  }
  l->lock_signal_posix();
	// TODO Would you like to join all the ptrheads
	//cerr<<"Waiting for threads to join"<<endl;
    	for(int i=0; i<workers.size(); i++) {
	  if(workers[i] != NULL) {
	    lock_var->lock_wait_posix();
	    pthread_cond_broadcast(&work);
	    lock_var->lock_signal_posix();
	    int res = pthread_join(id[i], NULL);
	    if(res) cerr<<"Could not join with thread!"<<endl;
	  }
	  workers[i] = NULL;
	}
}

    /**
     * Attempts to stop all actively executing tasks, and halts the
     * processing of waiting tasks.
     */
void Pool::shutdownNow() {
        PosixLock *l = this->lock_var;
        l->lock_wait_posix();
        //tryTerminate 
        l->lock_signal_posix();
				// TODO Would you like to join all the ptrheads
}

// RAJ:: timeout is used as it is without conversion
bool Pool::awaitTermination(const struct timespec *timeout) {
        
        PosixLock *l = this->lock_var;
        l->lock_wait_posix();
        
        for (;;) {
                if (runState == TERMINATED)
                    return true;
                if (timeout->tv_nsec <= 0)
                    return false;
                // nanos = termination.awaitNanos(timeout); // TODO wait for timeout nanos RAJ
								pthread_cond_timedwait(&termination,&this->lock_var->posix_lock_var,timeout);

            }
        l->lock_signal_posix();
}
// Internal methods that may be invoked by workers

bool Pool::isStopped() const {
     return runState >= STOP;
}

 /**
  * Return the workers array; needed for random-steal by Workers.
  */
/*
vector<Worker *> &Pool::getWorkers() const {
     return &workers;
}
*/
 
void Pool::submit(Job *job) {
    	//startTime = System.nanoTime();
        addJob(job);
}
void Pool::invoke(Job *job) {
    	submit(job);
    	//try {
    	job->waitForCompletion();
		//} catch (InterruptedException z) {} // TODO
}

/**
     * Enqueue an externally submitted task
     */
void Pool::addJob(Job *job) {
        PosixLock *l = this->lock_var;
        bool ok;
        l->lock_wait_posix();
        
        if (ok = (runState == EXECUTING)) {
                jobs->add(job);
                //work.signalAll();
                pthread_cond_broadcast(&work);
        }
        
        l->lock_signal_posix();
        
        if (!ok) { abort(); }
            //throw RejectedExecutionException();
}

/**
     * Returns the total number of tasks stolen from one thread's work
     * queue by another. This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steals.
     */
long Pool::getStealCount() const {
        long sum = 0;
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL) 
                sum += t->stealCount;
        }
        return sum;
}
    /**
     * Returns the total number of steal attempts made by all workers.
     * This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steal attempts.
     */
long Pool::getStealAttempts() const {
    	long sum = 0;
      for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL) 
                sum += t->stealAttempts;
      }
      return sum;
}
    /**
     * Termination callback from dying worker.
     */
void Pool::workerTerminated(Worker *r, int index) {
  /*sriramk: Not thought through the semantics in C++. The new in this
    function does not have a corresponding delete either. So
    aborting. */
  assert(0); 

        PosixLock *l = this->lock_var;
        l->lock_wait_posix();
        
        if (runState >= STOP) {
        	if (--runningWorkers <= 0) {
        		runState = TERMINATED;
						MEM_BARRIER(); // TODO RAj
                    //termination.signalAll();
        		pthread_cond_broadcast(&termination);
            }
        }
        else if (index >= 0 && index < workers.size() &&
                     workers[index] == r) {
        	abort();
        	/* // TODO Confirm from Vijay
            Worker *replacement = new Worker(this, index);
			assert(replacement != NULL);
            if (ueh != NULL)
                    replacement.setUncaughtExceptionHandler(ueh);
                workers[index] = replacement;
                replacement.start();
                */
         }
        
         l->lock_signal_posix();
}

void Pool::jobCompleted() {
        PosixLock *l = this->lock_var;
        l->lock_wait_posix();
        
        if (--activeJobs <= 0 && runState == SHUTDOWN && jobs->isEmpty())
                tryTerminate();
        
        l->lock_signal_posix();
        
}
void Pool::tryTerminate() const {
    	// do nothing.
}
Closure *Pool::getJob() {
        
  lock_var->lock_wait_posix();

  Job *task = jobs->poll();
  if (task == NULL && activeJobs == 0) {
    //work.await(); // TODO RAJ
    //cerr<<"waiting on condition"<<endl;
    pthread_cond_wait(&work,&(this->lock_var->posix_lock_var));
    //cerr<<"woke up from condition"<<endl;
    task=jobs->poll();
  }
  if (task != NULL) {
    ++ activeJobs;
    currentJob = task;
    MEM_BARRIER(); // Let the update be visible to everyone -- TODO RAJ
    //work.signalAll(); // TODO RAJ
    pthread_cond_broadcast(&work);
  }
  //} catch(InterruptedException e) { return NULL; }
  lock_var->lock_signal_posix();
  return task;
}

/*support for area-specific data */
void Pool::allocateWorkerArea(int size /*in bytes*/) {
  for(int i=0; i<workers.size(); i++) {
    while(workers[i] == NULL)
      sched_yield();
    assert(workers[i] != NULL);
    if(workers[i] != NULL) {
      void *area = malloc(size);
      assert(area != NULL);
      bzero(area, size);
      workers[i]->setArea(area);
      //cerr<<"Pool. set area="<<area<<" for worker id="<<i<<endl;
    }
  }
  MEM_BARRIER();
}

/*-------------------JobQueue--------------------*/

JobQueue::JobQueue() {
	elements.resize(INITIAL_JOBQUEUE_CAPACITY);
	head = tail = 0;
}
JobQueue::~JobQueue() {
  elements.clear();
}

bool JobQueue::isEmpty() const {
	return head == tail;
}

void JobQueue::add(Job *e) {
        elements[tail] = e;
        if ( (tail = (tail + 1) & (elements.size() - 1)) == head)
            doubleCapacity();
}

Job *JobQueue::poll() {
        int h = head;
        Job *result = elements[h]; 
        if (result != NULL) {
            elements[h] = NULL;
            head = (h + 1) & (elements.size() - 1);
        }
        return result;
}
   
void JobQueue::doubleCapacity() {
				int i;
        int p = head;
        int n = elements.size();
        int r = n - p; 
        int newCapacity = n << 1;
        if (newCapacity < 0) { abort();} // TODO throw exception if needed }
        elements.resize(newCapacity); 

				// TODO RAJ: please optimize this code later
				vector<Job *> tmp(INITIAL_JOBQUEUE_CAPACITY);
        for(i=p; i<n; i++) tmp[i-p]=elements[i];
				for(i=0;i<p;i++) tmp[r+i] = elements[i];
				for(i=0;i<n;i++) elements[i] = tmp[i];
				tmp.clear();
        head = 0;
        tail = n;
}
bool JobQueue::isQuiescent(vector<Worker *> &workers) const{
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL && t->isActive())
                return false;
        }
        return true;
}



